In [9]:
import pandas as pd
import joblib

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier


In [10]:
# Load Dataset
df = pd.read_csv("../datasets/clean_190k_dataset.csv")

print("Dataset Loaded Successfully!")
print(df.shape)

Dataset Loaded Successfully!
(189647, 378)


In [11]:
# Reduce dataset size to avoid MemoryError
df = df.sample(frac=0.25, random_state=42)

print("Sampled Dataset Shape:")
print(df.shape)

Sampled Dataset Shape:
(47412, 378)


In [12]:
X = df.drop("diseases", axis=1)

y = df["diseases"]

print("Features:", X.shape)
print("Target:", y.shape)

Features: (47412, 377)
Target: (47412,)


In [13]:
label_encoder = LabelEncoder()

y_encoded = label_encoder.fit_transform(y)

print("Number of Disease Classes:", len(label_encoder.classes_))

Number of Disease Classes: 700


In [14]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y_encoded,
    test_size=0.2,
    random_state=42
)

print("Training Set:", X_train.shape)
print("Testing Set:", X_test.shape)

Training Set: (37929, 377)
Testing Set: (9483, 377)


In [15]:
rf_model = RandomForestClassifier(
    n_estimators=30,
    max_depth=15,
    max_features="sqrt",
    random_state=42,
    n_jobs=1
)

In [16]:
print("Training Started...")

rf_model.fit(X_train, y_train)

print("Model Trained Successfully!")

Training Started...
Model Trained Successfully!


In [17]:
joblib.dump(rf_model, "../models/random_forest_model.pkl")

joblib.dump(label_encoder, "../models/label_encoder.pkl")

print("Model Saved Successfully!")

Model Saved Successfully!


## Hardware Limitation

The Random Forest model could not be trained successfully on the complete
dataset (189,647 records and 773 disease classes) using the available
hardware (8 GB RAM).

During training, a MemoryError occurred because Random Forest constructs
multiple decision trees simultaneously, requiring more memory than was
available.

The implementation is correct and can be trained successfully on a machine
with higher RAM (16 GB or more) or on cloud computing platforms.